### PR TITLE
526 data translations

### PR DESCRIPTION
**Pull request by SMLuthi**
When a user submits a 526EZ form for submission certain data translations and additions have to take place to make it valid. Below is a list of specifics:

- [x] add `claimantCertification` field
- [x] add `applicationExpirationDate` field
- [x] add `directDeposit` field
- [x] translate `treatments` field
- [x] translate `serviceInformation` and its subfields `confinements`, `alternateNames`, and `reservesNationalGuardService`
- [x] translate `veteran' and its subfields `mailingAddress` and `forwardingAddress`

Notes:

The calculations for the `applicationExpirationDate` was drawn from here: https://github.com/department-of-veterans-affairs/vets.gov-team/blob/master/Products/Disability%20Compensation/engineering/EVSS/swagger_q%26a.md

Currently the conversion is taking place within the controller submit method. Once this is moved back to being an asynchronous call the conversion can be moved within the worker to take advantage of it.

### CODE REVIEW

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+            'accountNumber' => account&.account_number,
+            'routingNumber' => account&.financial_institution_routing_number,
+            'bankName' => account&.financial_institution_name
+          }
+        end
+      end
+
+      def convert_mailing_address(address)
+        pciu_address = { 'country' => address['country'],
+                         'addressLine1' => address['addressLine1'],
```
Comment (cehsu) [inline lib/evss/disability_compensation_form/data_translation.rb]: The naming here was [discussed](https://github.com/department-of-veterans-affairs/vets-website/pull/7846/files#r194169990) recently, but I'm not sure whether there was consensus on if changing this was of interest.

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+module EVSS
+  module DisabilityCompensationForm
+    class DataTranslation
+      def initialize(user, form_content)
+        @user = user
+        @form_content = form_content
+      end
+
+      def convert
```
Comment (rlbaker) [inline lib/evss/disability_compensation_form/data_translation.rb]: Definitely a nitpick, but would `translate` would be a better name for the main method being used of this `DataTranslation` class?

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+
+      def application_create_date
+        # Application create date is the date the user began their application
+        @acd ||= InProgressForm.where(form_id: VA526ez::FORM_ID, user_uuid: @user.uuid)
+                               .first.created_at
+      end
+
+      def rad_date
+        # retrieve the most recent 'Return from Active Duty' Date
+        @rd ||= Time.zone.parse(@user.military_information.service_episodes_by_date.first&.end_date.to_s)
```
Comment (rlbaker) [inline lib/evss/disability_compensation_form/data_translation.rb]: I would split this line up for readability

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+      end
+
+      def itf
```
Comment (kreek) [inline lib/evss/disability_compensation_form/data_translation.rb]: What happens if `IntentToFile::Service.get_active` fails? As you mention in the controller the submit process should go into an async job (or multiple chained jobs) but we can't do that until we know that we won't get a validation error from EVSS (i.e. all the validation is on our side)

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+          'title10Activation' => reserves_service_info['title10Activation'],
+          'obligationTermOfServiceFromDate' => reserves_service_info['obligationTermOfServiceDateRange']['from'],
+          'obligationTermOfServiceToDate' => reserves_service_info['obligationTermOfServiceDateRange']['to'],
+          'unitName' => reserves_service_info['unitName'],
+          'unitPhone' => split_phone_number(reserves_service_info['unitPhone']),
+          'inactiveDutyTrainingPay' => reserves_service_info['inactiveDutyTrainingPay']
+        }.compact
+      end
+
+      def convert_veteran
```
Comment (kreek) [inline lib/evss/disability_compensation_form/data_translation.rb]: although it's all veteran data could there be separate methods for phone, address, and homelessness?

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+
+        homeless_data = veteran['homelessness']
+        if homeless_data['isHomeless'].present?
+          veteran['homelessness'] = convert_homelessness(homeless_data['pointOfContact'])
+        else
+          veteran.delete('homelessness')
+        end
+      end
+
+      def convert_homelessness(point_of_contact)
```
Comment (kreek) [inline lib/evss/disability_compensation_form/data_translation.rb]: if you do split the methods up could all the homelessness logic be in one method?
```ruby
def convert_homelessness
  data = veteran['homelessness']
  return veteran.delete('homelessness') unless data['isHomeless'].present?
  return veteran['homelessness'] = { 'hasPointOfContact' => false } unless data['pointOfContact'].present?
  veteran['homelessness'] = {
    'hasPointOfContact' => true,
    'pointOfContact' => {
      'pointOfContactName' => data.dig('pointOfContact', 'pointOfContactName'),
      'primaryPhone' => split_phone_number(data.dig('pointOfContact', 'primaryPhone'))
    }
  }
end
```

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            'primaryPhone' => split_phone_number(point_of_contact['primaryPhone'])
+          }
+        else
+          homelessness['hasPointOfContact'] = false
+        end
+
+        homelessness
+      end
+
+      def service_branch(service_branch)
```
Comment (kreek) [inline lib/evss/disability_compensation_form/data_translation.rb]: could have a guard clause here too
```ruby
def service_branch(service_branch)
  return 'National Oceanic &amp; Atmospheric Administration' if service_branch == 'NOAA'
  service_branch
end
```

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+      def convert_treatments
+        form['treatments'].map! do |treatment|
+          treatment['center'] = {
+            'name' => treatment['treatmentCenterName'],
+            'type' => treatment['treatmentCenterType']
+          }
+          treatment['center'].merge!(treatment['treatmentCenterAddress'])
+          treatment['startDate'] = treatment['treatmentDateRange']['from']
+          treatment['endDate'] = treatment['treatmentDateRange']['to']
+          treatment.delete('treatmentCenterName')
```
Comment (kreek) [inline lib/evss/disability_compensation_form/data_translation.rb]: return hash.except from the block and you could replace the deletes
```ruby
treatment.except('treatmentCenterName', 'treatmentDateRange', 'treatmentCenterAddress', 'treatmentCenterType')
```

File: lib/evss/disability_compensation_form/data_translation.rb [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+          pciu_address['militaryPostOfficeTypeCode'] = address['city']
+          pciu_address['militaryStateCode'] = address['state']
+        when 'INTERNATIONAL'
+          pciu_address['city'] = address['city']
+        end
+
+        pciu_address.compact
+      end
+
+      def get_address_type(address)
```
[the remark this answers is not in the document]
Comment (SMLuthi) [reply lib/evss/disability_compensation_form/data_translation.rb]: This has been updated as requested.

**Pull request merged**